core/domain/exp_services_test.py

```diff
@@ -1062,6 +1084,304 @@ def test_save_and_retrieve_exploration_summary(self):
+            }),
+            exp_domain.ExplorationChange({
+                'cmd': exp_domain.CMD_EDIT_STATE_PROPERTY,
+                'state_name': exploration.init_state_name,
+                'property_name': (
+                    exp_domain.STATE_PROPERTY_INTERACTION_CUST_ARGS),
+                'new_value': {
+                    'language': {
+                        'value': 'python'
+                    },
+                    'placeholder': {
+                        'value': '# Type your code here.'
+                    },
+                    'preCode': {
+                        'value': ''
+                    },
+                    'postCode': {
+                        'value': ''
+                    }
+                    }
+                })]
+        exp_services.update_exploration(
+            self.owner_id, self.EXP_0_ID, change_list, 'Changed to CodeRepl')
+        updated_exploration = exp_fetchers.get_exploration_by_id(self.EXP_0_ID)
+        errors = exp_services.validate_exploration_for_story(
+            updated_exploration, False)
+        self.assertEqual(len(errors), 1)
+        self.assertEqual(errors[0], error_string)
+        with self.assertRaisesRegex(
+            utils.ValidationError, error_string):
+            exp_services.validate_exploration_for_story(
+                updated_exploration, True)
+
+    def test_validation_fail_for_end_exploration(self):
+        exploration = self.save_new_valid_exploration(
+            self.EXP_0_ID, self.owner_id, correctness_feedback_enabled=True)
+        error_string = (
+            'Exploration with ID: %s contains exploration '
+            'recommendations in its EndExploration interaction.'
+            % (exploration.id)
+        )
+        change_list = [
+            exp_domain.ExplorationChange({
+                'cmd': exp_domain.CMD_EDIT_STATE_PROPERTY,
+                'state_name': exploration.init_state_name,
+                'property_name': exp_domain.STATE_PROPERTY_INTERACTION_ID,
+                'new_value': 'EndExploration'
+                }),
+            exp_domain.ExplorationChange({
+                'cmd': exp_domain.CMD_EDIT_STATE_PROPERTY,
+                'state_name': exploration.init_state_name,
+                'property_name': (
+                    exp_domain.STATE_PROPERTY_INTERACTION_CUST_ARGS),
+                'new_value': {
+                    'recommendedExplorationIds': {
+                        'value': [
+                            'EXP_1',
+                            'EXP_2'
+                            ]
+                        }
+                    }
+                }),
+            exp_domain.ExplorationChange({
+                'cmd': exp_domain.CMD_EDIT_STATE_PROPERTY,
+                'property_name': (
+                    exp_domain.STATE_PROPERTY_INTERACTION_DEFAULT_OUTCOME),
+                'state_name': exploration.init_state_name,
+                'new_value': None}
+            )]
```

Comment:
It's confusing to put the `}` and `)` on different lines when the associated `{` and `(` are on the same line. I'd recommend this instead:

```python
[
    ...
    exp_domain.ExplorationChange({
        ...
        'new_value': None,
    })
]
``` 

Notice how this approach uses the indentation to make clear which closing brackets/braces/parentheses go with which opening brackets/braces/parentheses